### ISSUE TITLE
Use UNORDERED IBC channels

### ISSUE THREAD

**Issue by AdityaSripal** [informalsystems/cross-chain-validation#12]
The spec for CC1 mentions that cross-chain validation will use ORDERED channels: "We assume that the IBC channel is ordered"

ORDERED channels is not a great choice for this use-case, since a single timeout in the packet timeline will cause the channel to close and all later packets to not be receivable: 

https://github.com/cosmos/cosmos-sdk/blob/master/docs/ibc/overview.md#receipts-and-timeouts

The spec already mentions a queue for packets. This can be used to enforce packet-processing order on the application side.

The queue can be a queue ordered by `seqNum`, the queue can be filled in from the UNORDERED channel out of order, and the queue adds each packet in the order of the `seqNum`. The application also keeps track of the latest seqNum `latestProcessed` that has been dequeued and processed.

At each `Endblock`, the queue can be checked to see if `latestProcessed+1` is at the head of the queue. If it is, then we can dequeue and process the validator set changes. We then write the acknowledgement for this packet (using async acknowledgements) and increment the `latestProcessed` sequence.

If the queue has not received the next packet in the sequence yet, it can continue enqueuing packets out of order while we wait for the next packet to arrive. This also allows us to timeout packets on the channel safely, since the parent chain can receive the timeout and simply resend the packet with the same application-defined `seqNum`.

Note: The `seqNum` here is a separate sequence tracked by the application that is different from the IBC packet sequence maintained and tracked by IBC.

cc: @cwgoes

**Comment by cwgoes** [comment]
Hmm, I don't really understand the advance of this approach over using a long timeout window (e.g. close to the unbonding period) - if there's a long disconnection between the chains, timed-out packets can't be resent anyways (until it closes), and if the unbonding period is exceeded the light client updates would fail anyways. Is there an advantage here?

**Comment by josef-widder** [comment]
I agree with @cwgoes: we should use the ordered channels of IBC (with long timeouts), rather than implementing ordered channels on top of unordered ones. 

After all, we are not dealing with "packet loss" that the participants cannot control here. My understanding is that the validators that are involved in a validator set change have the incentive to get the update on the baby chain: either (in case they leave the validator set) to get their stake back a.s.a.p or (in case they enter the validator set) to start validating and earning rewards. So they have an interest in relaying.

**Comment by josef-widder** [comment]
@jovankomatovic might be interested in this discussion...
